### ISSUE TITLE
Error connecting to database on wp rewrite structure

### ISSUE THREAD

**Issue by diamondfish** [wp-cli/wp-cli#3681]
I've got everything to work with WP-CLI on Windows running XAMPP. Or so I thought. When I try to execute the following command:

`wp rewrite structure '/%postname%/'`

I get the following error:

> Error: Error establishing a database connection. This either means that the username and password information in your wp-config.php file is incorrect or we canÔÇÖt contact the database server at localhost. This could mean your hostÔÇÖs database server is down.

I've read that the PHP used needs to be correct (a usual explanation for this problem). Running  wp --info yields the following results.

```
$ wp --info
PHP binary:     C:\xampp\xampp-5.6.24\php\php.exe
PHP version:    5.6.24
php.ini used:   C:\xampp\xampp-5.6.24\php\php.ini
WP-CLI root dir:        C:\lib\wp-cli
WP-CLI packages dir:
WP-CLI global config:
WP-CLI project config:
WP-CLI version: 1.0.0
```

I think that this looks correct. I can access the db from a browser, use WP normally, and even perform database commands using wp-cli. But for some reason, the rewrite command gives an error (as stated above). The fun thing is that it actually changes the permalink option, but I'm afraid that since it gives an error some other action is not performed correctly.

Changing the credentials in wp-config.php makes me unable to run db commands, so clearly it works at some level.

What might be wrong? I'm pretty clueless here!

The database is set up by using wp-cli and is accessed using the root account. I've also tried using the .phar version.

**Comment by diamondfish** [comment]
Missclicked enter and the issue was posted before I got to enter the information. Sorry about that. But to my problem:

I've got everything to work with WP-CLI on Windows running XAMPP. Or so I thought. When I try to execute the following command:

`wp rewrite structure '/%postname%/'`

I get the following error:

> Error: Error establishing a database connection. This either means that the username and password information in your wp-config.php file is incorrect or we canÔÇÖt contact the database server at localhost. This could mean your hostÔÇÖs database server is down.

I've read that the PHP used needs to be correct (a usual explanation for this problem). Running  wp --info yields the following results.

```
$ wp --info
PHP binary:     C:\xampp\xampp-5.6.24\php\php.exe
PHP version:    5.6.24
php.ini used:   C:\xampp\xampp-5.6.24\php\php.ini
WP-CLI root dir:        C:\lib\wp-cli
WP-CLI packages dir:
WP-CLI global config:
WP-CLI project config:
WP-CLI version: 1.0.0
```

I think that this looks correct. I can access the db from a browser, use WP normally, and even perform database commands using wp-cli. But for some reason, the rewrite command gives an error (as stated above). The fun thing is that it actually changes the permalink option, but I'm afraid that since it gives an error some other action is not performed correctly.

Changing the credentials in wp-config.php makes me unable to run db commands, so clearly it works at some level.

What might be wrong? I'm pretty clueless here!

The database is set up by using wp-cli and is accessed using the root account.

**Comment by danielbachhuber** [comment]
It may be related to #3640

How are your database credentials set in `wp-config.php` ?

**Comment by diamondfish** [comment]
I have read that issue and was not able to find a solution.

I create the `wp-config.php` file by using this command:
`wp core config --dbname="$dbName" --dbuser="$dbUser" --dbpass="$dbPass" --locale="$dbLocale" --dbcharset="$dbCharset" --dbcollate="$dbCollate"`

The wp-config file reads like this:

```
// ** MySQL settings ** //
/** The name of the database for WordPress */
define( 'DB_NAME', 'wp_project3' );

/** MySQL database username */
define( 'DB_USER', 'root' );

/** MySQL database password */
define( 'DB_PASSWORD', 'mypassword' );

/** MySQL hostname */
define( 'DB_HOST', 'localhost' );

/** Database Charset to use in creating database tables. */
define( 'DB_CHARSET', 'utf8' );

/** The Database Collate type. Don't change this if in doubt. */
define( 'DB_COLLATE', 'utf8_general_ci' );
```

I have tried changing to 127.0.0.1 as host, edit MySQL settings in `my.ini`, creating another database account etc, but with no effect. And I can run other db commands like this:

```
project3 $ wp db query 'SHOW TABLES' --debug
Debug (bootstrap): No readable global config found (0.007s)
Debug (bootstrap): No project config found (0.01s)
Debug (bootstrap): No package autoload found to load. (0.097s)
Debug (bootstrap): ABSPATH defined: C:\www\project3/ (0.098s)
Debug (bootstrap): Running command: db query (0.101s)
+-----------------------+
| Tables_in_wp_project3 |
+-----------------------+
| wp_commentmeta        |
| wp_comments           |
| wp_links              |
| wp_options            |
| wp_postmeta           |
| wp_posts              |
| wp_term_relationships |
| wp_term_taxonomy      |
| wp_termmeta           |
| wp_terms              |
| wp_usermeta           |
| wp_users              |
+-----------------------+
```

The rewrite command with --debug added looks like this:

```
project3 $ wp rewrite structure '/%postname%/' --debug
Debug (bootstrap): No readable global config found (0.007s)
Debug (bootstrap): No project config found (0.008s)
Debug (bootstrap): No package autoload found to load. (0.097s)
Debug (bootstrap): ABSPATH defined: C:\www\project3/ (0.097s)
Debug (bootstrap): Begin WordPress load (0.098s)
Debug (bootstrap): wp-config.php path: C:\www\project3\wp-config.php (0.098s)
Debug (bootstrap): Loaded WordPress (1.277s)
Debug (bootstrap): Running command: rewrite structure (1.277s)
PHP Notice:  Undefined variable: key in phar://C:/lib/wp-cli-phar/wp-cli.phar/php/class-wp-cli.php on line 1003
Notice: Undefined variable: key in phar://C:/lib/wp-cli-phar/wp-cli.phar/php/class-wp-cli.php on line 1003
Debug (bootstrap): No readable global config found (0.007s)
Debug (bootstrap): No project config found (0.008s)
Debug (bootstrap): No package autoload found to load. (0.098s)
Debug (bootstrap): ABSPATH defined: C:\www\project3/ (0.098s)
Debug (bootstrap): Begin WordPress load (0.099s)
Debug (bootstrap): wp-config.php path: C:\www\project3\wp-config.php (0.099s)
Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or we canÔÇÖt contact the database server at `localhost`. This could mean your hostÔÇÖs database server is down.
```

I've also tried correcting the bug (Undefined variable: key in phar) that you fixed in an upcoming release, according to [this git issue](https://github.com/wp-cli/wp-cli/commit/214f47af35b9b0138b51b6ecac8a744e5410b34a).

**Comment by danielbachhuber** [comment]
Interesting.

The surface problem is that WP-CLI can't connect to the database within the spawned child process. I'm not sure why that'd be the case, though.

**Comment by diamondfish** [comment]
Is there anything I can force to make it work? Or at least test?

**Comment by danielbachhuber** [comment]
You can call `wp rewrite flush` after `wp rewrite structure`

What `my.ini` settings did you change?

**Comment by diamondfish** [comment]
If i use `wp rewrite structure`, which actually changes the option of permalinks in WP even though it gives error afterwards, followed by `wp rewrite flush`, will I get the same result as if the first command didn't give any error? If you understand what I mean. I'm just worried that some important action after the option is set in WP is missed due to the error.

`wp rewrite flush` gives me success message so it should work fine for me to do both, given that nothing is missed like I said.

The settings I played around with was (in various combinations):
```
[client] 
host="localhost"
host="127.0.0.1"
protocol=tcp
socket = ""

[mysqld]
socket=MySQL
enable-named-pipe=1

default-time-zone=SYSTEM
max_allowed_packet = 32M

bind-address="127.0.0.1" 
```

**Comment by danielbachhuber** [comment]
> If i use `wp rewrite structure`, which actually changes the option of permalinks in WP even though it gives error afterwards, followed by `wp rewrite flush`, will I get the same result as if the first command didn't give any error?

Yes. WP-CLI's `wp rewrite structure` calls `wp rewrite flush` under the hood, so they're functionally equivalent.

> The settings I played around with was (in various combinations):

Nothing obvious here.

Is there some amount of your system configuration that's served by environment variable?

**Comment by diamondfish** [comment]
That's nice. So then I can call both functions and getting full functionality even with the error message I assume.

I use the following added to my System Variable PATH:

```
C:\xampp\xampp-5.6.24\php
C:\xampp\xampp-5.6.24\mysql\bin
C:\xampp\xampp-5.6.24\apache\bin
C:\lib\wp-cli-phar
```

**Comment by danielbachhuber** [comment]
> I use the following added to my System Variable PATH:

Ah, that's the issue. `WP_CLI::runcommand()` needs to persist `$PATH` too. In your case, it's using the wrong MySQL binary.

**Comment by diamondfish** [comment]
That sounds interesting! Although, I'm not sure exactly what to change. Can you please provide a little bit more information on this?

**Comment by danielbachhuber** [comment]
> Although, I'm not sure exactly what to change. Can you please provide a little bit more information on this?

No worries, handling in #3683

**Comment by diamondfish** [comment]
I don't know if this helps you, but the only mysql*.exe files I have is within my `C:\xampp\xampp-5.6.24\mysql\bin\` folder.

**Comment by mkeerth** [comment]
@danielbachhuber : I'm getting the same exception "Error establishing connection". But i'm running this command "wp --allow-root rewrite structure" from inside php:7.2-fpm-alpine3.6 Docker container.
can you please help me?

**Comment by webdesignermall** [comment]
Getting same issue using MAMP typing in something simple like wp plugin list. I get PHP Warning:  

> mysqli_real_connect(): (HY000/2002): No such file or directory in /Applications/MAMP/htdocs/wordpress49/wp-includes/wp-db.php on line 1531
> Warning: mysqli_real_connect(): (HY000/2002): No such file or directory in /Applications/MAMP/htdocs/wordpress49/wp-includes/wp-db.php on line 1531
> Error: Error establishing a database connection. This either means that the username and password information in your `wp-config.php` file is incorrect or we can’t contact the database server at `localhost`. This could mean your host’s database server is down.

Can someone shed some light?

**Comment by danielbachhuber** [comment]
Hi folks,

GitHub issues are meant for enhancement requests and specific, reproducible bugs, not for general support questions. For support options, please review http://wp-cli.org/#support

**Comment by totten** [comment]
Tangential note for anyone debugging this error message... I tried `--debug` and it didn't yield enough info to see the problem. So I pulled up a debugger... and still had trouble seeing the problem.

The bit which finally helped: edit `wp-config.php` and set `WP_DEBUG` more directly. That yielded a lot more info about database issues.

(*For me, the problem turned out to have nothing to do with `wp-cli` or WordPress... the (old-ish) `mysqld` got into a weird state where it would give a good mysql connection to the first PHP/wp-cli process... and then give a bad connection for the second process. I just needed a way to see the errors...*)
